File: services/api-node/src/lexicon.js
import fs from 'node:fs'
import path from 'node:path'
import xlsx from 'xlsx'

export function defaultXlsxPaths(projectRoot) {
  return [
    path.join(projectRoot, '京师小学考纲.xlsx'),
    path.join(projectRoot, '京师初中考纲（不含小学）.xlsx'),
    path.join(projectRoot, '京师高中考纲（不含初中）.xlsx'),
  ]
}

export function inferStageFromFilename(name) {
  if (name.includes('高中')) return '高中'
  if (name.includes('初中')) return '初中'
  if (name.includes('小学')) return '小学'
  return '未知'
}

export function loadLexicon(projectRoot) {
  const out = []
  for (const p of defaultXlsxPaths(projectRoot)) {
    if (!fs.existsSync(p)) {
      console.log('Missing file:', p)
      continue
    }
    const stage = inferStageFromFilename(path.basename(p))
    const items = loadXlsx(p, stage)
    console.log(`Loaded ${items.length} items for stage ${stage} from ${path.basename(p)}`)
    out.push(...items)
  }
  return assignLevels(dedupItems(out), projectRoot)
}

function loadXlsx(filePath, stage) {
  const wb = xlsx.readFile(filePath, { cellDates: false })
  const out = []
  for (const sheetName of wb.SheetNames) {
    const ws = wb.Sheets[sheetName]
    if (!ws) continue
    const rows = xlsx.utils.sheet_to_json(ws, { defval: null })
    if (!rows.length) continue
    const columns = normalizeColumns(Object.keys(rows[0] || {}))
    const colWord = pickCol(columns, ['word', '单词', '词汇', '英文'])
    if (!colWord) {
      console.log('Skipping sheet', sheetName, 'because no word column found in', columns)
      continue
    }
    const colMeaning = pickCol(columns, [
      'meaning_zh',
      '译文',
      '中文',
      '释义',
      '中文释义',
      '翻译',
      '意义',
    ])
    const colLevel = pickCol(columns, ['level', '级', '等级', '层级'])
    const colPos = pickCol(columns, ['pos', '词性'])
    const colPhonetic = pickCol(columns, ['phonetic', '音标', '音标(英)', '音标(美)'])

    for (const raw of rows) {
      const row = normalizeRowKeys(raw)
      const w0 = asStr(row[colWord])
      if (!w0) continue
      const word = cleanWord(w0)
      const meaning = colMeaning ? asStr(row[colMeaning]) : ''
      const level = colLevel ? normalizeLevel(asStr(row[colLevel])) : ''
      const phonetic = colPhonetic ? normalizePhonetic(asStr(row[colPhonetic])) : ''
      const pos =
        (colPos ? normalizePos(asStr(row[colPos])) : '') ||
        (meaning ? extractPosFromMeaning(meaning) : '')
      out.push({
        word,
        meaning_zh: meaning || null,
        stage,
        level: level || null,
        pos: pos || null,
        phonetic: phonetic || null,
      })
    }
  }
  return out
}

function normalizeColumns(cols) {
  return cols.map((c) =>
    asStr(c).replace(/\r|\n/g, '').trim().replace(/\s+/g, ''),
  )
}

function normalizeRowKeys(row) {
  const out = {}
  for (const [k, v] of Object.entries(row)) {
    const nk = asStr(k).replace(/\r|\n/g, '').trim().replace(/\s+/g, '')
    out[nk] = v
  }
  return out
}

function pickCol(cols, candidates) {
  for (const c of candidates) {
    const exact = cols.find((col) => col.toLowerCase() === c.toLowerCase())
    if (exact) return exact
  }
  for (const col of cols) {
    for (const c of candidates) {
      if (col.toLowerCase().includes(c.toLowerCase())) return col
    }
  }
  return null
}

function asStr(v) {
  if (v == null) return ''
  const s = String(v).trim()
  if (!s) return ''
  const low = s.toLowerCase()
  if (low === 'nan' || low === 'none' || low === 'null') return ''
  return s
}

function cleanWord(v) {
  return v.trim()
}

function normalizeLevel(v) {
  const s = v.trim()
  if (!s) return s
  const s2 = s.replace('级别', '级').replace('等级', '级')
  const m = /(小学|初中|高中)?\s*(\d+)\s*级/.exec(s2)
  if (m) {
    const n = Number(m[2])
    const prefix = m[1]
    return prefix ? `${prefix} ${n}级` : `${n}级`
  }
  const m2 = /(\d+)/.exec(s2)
  if (m2) return `${Number(m2[1])}级`
  return s2
}

function normalizePos(v) {
  const s = v.trim().toLowerCase()
  return s
    .replaceAll('名词', 'n')
    .replaceAll('动词', 'v')
    .replaceAll('形容词', 'adj')
    .replaceAll('副词', 'adv')
}

function normalizePhonetic(v) {
  const s = v.trim()
  if (!s) return s
  if (s.startsWith('/') && s.endsWith('/') && s.length >= 2) return s.slice(1, -1)
  return s
}

function extractPosFromMeaning(meaning) {
  const m = /^\s*([a-zA-Z]{1,5})\./.exec(meaning)
  if (!m) return ''
  const raw = m[1].toLowerCase()
  const known = new Set([
    'n',
    'v',
    'vi',
    'vt',
    'adj',
    'adv',
    'prep',
    'conj',
    'pron',
    'num',
    'art',
    'interj',
    'aux',
  ])
  if (!known.has(raw)) return raw
  return raw
}

function dedupItems(items) {
  const seen = new Set()
  const out = []
  for (const it of items) {
    const k = `${it.stage}::${it.word.toLowerCase()}`
    if (seen.has(k)) continue
    seen.add(k)
    out.push(it)
  }
  return out
}

function assignLevels(items, projectRoot) {
  const txtPath = path.join(projectRoot, 'services/api-node/data/freq-10k.txt')
  const fMap = new Map()
  try {
    const content = fs.readFileSync(txtPath, 'utf8')
    const lines = content.split('\n')
    for (let i = 0; i < lines.length; i++) {
      const w = lines[i].trim().toLowerCase()
      if (w && !fMap.has(w)) {
        fMap.set(w, i + 1)
      }
    }
  } catch (e) {
    console.error('Failed to load freq map', e)
  }

  function getRank(w) {
    const wLow = w.toLowerCase()
    if (fMap.has(wLow)) return fMap.get(wLow)
    return 10000 + wLow.length * 100
  }

  function getCEFR(rank) {
    if (rank <= 1500) return 'A1 (入门)'
    if (rank <= 3000) return 'A2 (初级)'
    if (rank <= 5000) return 'B1 (中级)'
    if (rank <= 8000) return 'B2 (中高级)'
    if (rank <= 12000) return 'C1 (高级)'
    return 'C2 (精通)'
  }

  for (const it of items) {
    const r = getRank(it.word)
    it.level = getCEFR(r)
  }
  return items
}

File: services/api/app/lexicon.py
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import pandas as pd


@dataclass(frozen=True)
class WordItem:
    word: str
    meaning_zh: str | None
    stage: str
    level: str | None
    pos: str | None


def default_xlsx_paths(project_root: Path) -> list[Path]:
    return [
        project_root / "京师小学考纲.xlsx",
        project_root / "京师初中考纲（不含小学）.xlsx",
        project_root / "京师高中考纲（不含初中）.xlsx",
    ]


def load_lexicon(project_root: Path) -> list[WordItem]:
    items: list[WordItem] = []
    for p in default_xlsx_paths(project_root):
        if not p.exists():
            continue
        stage = infer_stage_from_filename(p.name)
        items.extend(load_xlsx(p, stage))
    return dedup_items(items)


def infer_stage_from_filename(name: str) -> str:
    if "小学" in name:
        return "小学"
    if "初中" in name:
        return "初中"
    if "高中" in name:
        return "高中"
    return "未知"


def load_xlsx(path: Path, stage: str) -> list[WordItem]:
    sheets: dict[str, Any] = pd.read_excel(path, sheet_name=None)
    out: list[WordItem] = []
    for _, df in sheets.items():
        if df is None or df.empty:
            continue
        df = normalize_columns(df)
        col_word = pick_col(df.columns, ["word", "单词", "词汇", "英文"])
        if not col_word:
            continue
        col_meaning = pick_col(df.columns, ["meaning_zh", "中文", "释义", "中文释义", "翻译", "意义"])
        col_level = pick_col(df.columns, ["level", "级", "等级", "层级"])
        col_pos = pick_col(df.columns, ["pos", "词性"])

        for _, row in df.iterrows():
            w = as_str(row.get(col_word))
            if not w:
                continue
            w = clean_word(w)
            meaning = as_str(row.get(col_meaning)) if col_meaning else None
            level = as_str(row.get(col_level)) if col_level else None
            pos = as_str(row.get(col_pos)) if col_pos else None
            out.append(
                WordItem(
                    word=w,
                    meaning_zh=meaning or None,
                    stage=stage,
                    level=normalize_level(level) if level else None,
                    pos=normalize_pos(pos) if pos else None,
                )
            )
    return out


def normalize_columns(df: pd.DataFrame) -> pd.DataFrame:
    def n(s: Any) -> str:
        v = as_str(s)
        v = v.replace("\n", "").replace("\r", "").strip()
        v = re.sub(r"\s+", "", v)
        return v

    df = df.copy()
    df.columns = [n(c) for c in df.columns]
    return df


def pick_col(cols: list[str], candidates: list[str]) -> str | None:
    for c in candidates:
        if c in cols:
            return c
    for c in cols:
        for k in candidates:
            if k in c:
                return c
    return None


def as_str(v: Any) -> str:
    if v is None:
        return ""
    if isinstance(v, float) and pd.isna(v):
        return ""
    s = str(v).strip()
    if s.lower() in {"nan", "none", "null"}:
        return ""
    return s


_word_re = re.compile(r"[A-Za-z][A-Za-z\\-']*")


def clean_word(v: str) -> str:
    m = _word_re.search(v)
    if not m:
        return v.strip()
    return m.group(0).strip()


def normalize_level(v: str) -> str:
    s = v.strip()
    if not s:
        return s
    s = s.replace("级别", "级").replace("等级", "级")
    m = re.search(r"(小学|初中|高中)?\\s*(\\d+)\\s*级", s)
    if m:
        n = int(m.group(2))
        prefix = m.group(1)
        if prefix:
            return f"{prefix} {n}级"
        return f"{n}级"
    m2 = re.search(r"(\\d+)", s)
    if m2:
        return f"{int(m2.group(1))}级"
    return s


def normalize_pos(v: str) -> str:
    s = v.strip().lower()
    s = s.replace("名词", "n").replace("动词", "v").replace("形容词", "adj").replace("副词", "adv")
    return s


def dedup_items(items: list[WordItem]) -> list[WordItem]:
    seen: set[tuple[str, str]] = set()
    out: list[WordItem] = []
    for it in items:
        k = (it.stage, it.word.lower())
        if k in seen:
            continue
        seen.add(k)
        out.append(it)
    return out

